File: PAT_Basic_Level_by_go/1022 D进制的A+B/1022 D进制的A+B.go
/*
 * 题目：1022 D 进制的 A+B
 * 实现原理：先计算 A+B 的和，然后用除 D 取余法将和转换为 D 进制数，
 * 注意处理和为 0 的特殊情况，最后按高位到低位输出。
 */
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// 将 A+B 的结果转换为 D 进制输出。
func main() {
	scanner := bufio.NewScanner(os.Stdin)
	if !scanner.Scan() {
		return
	}
	f := strings.Fields(scanner.Text())
	a, _ := strconv.Atoi(f[0])
	b, _ := strconv.Atoi(f[1])
	d, _ := strconv.Atoi(f[2])
	sum := a + b
	if sum == 0 {
		fmt.Println("0") // 和为零时直接输出 0
		return
	}
	var digits []int
	for sum > 0 {
		// 每次取余数并插入到切片头部，保证高位在前
		digits = append([]int{sum % d}, digits...)
		sum /= d
	}
	var b2 strings.Builder
	for _, x := range digits {
		b2.WriteString(strconv.Itoa(x))
	}
	fmt.Println(b2.String())
}

File: PAT_Basic_Level_by_c/1022 D进制的A+B/1022 D进制的A+B.c
/*
 * 题目：1022 D进制的A+B
 * 实现原理：
 *   计算两个整数A和B的和，并以D进制输出结果。
 *   算法步骤：
 *   1. 计算A+B的和
 *   2. 特殊处理：若和为0，直接输出0
 *   3. 使用除基取余法将十进制转换为D进制：
 *      - 不断用sum除以D，取余数
 *      - 余数逆序排列即为D进制结果
 *   4. 从高位到低位输出结果
 *   使用long long类型避免溢出。
 */
#include <stdio.h>

int main() {
    long long a, b;
    int d;                          // 目标进制
    scanf("%lld %lld %d", &a, &b, &d);
    long long sum = a + b;
    if (sum == 0) {                 // 和为 0 时直接输出 0
        printf("0\n");
        return 0;
    }
    int digits[32], len = 0;        // digits 逆序存放 D 进制各位，len 为位数
    while (sum > 0) {               // 除基取余法：反复除以 d 取余数
        digits[len++] = sum % d;    // 记录余数
        sum /= d;
    }
    for (int i = len - 1; i >= 0; i--) {   // 逆序输出，即从高位到低位
        printf("%d", digits[i]);
    }
    printf("\n");
    return 0;
}